flink-ml-lib/src/main/java/org/apache/flink/ml/common/datastream/AllReduceUtils.java

```diff
@@ -0,0 +1,286 @@
+                                        BasicTypeInfo.INT_TYPE_INFO,
+                                        PrimitiveArrayTypeInfo.DOUBLE_PRIMITIVE_ARRAY_TYPE_INFO),
+                                new AllReduceSum())
+                        .name("all-reduce-sum");
+
+        return allReduceSum
+                .partitionCustom((taskIdx, numPartitions) -> taskIdx % numPartitions, x -> x.f0)
+                .transform(
+                        "all-reduce-recv", TypeInformation.of(double[].class), new AllReduceRecv())
+                .name("all-reduce-recv");
+    }
+
+    /**
+     * Splits each double array into multiple chunks and send each chunk to the corresponding
+     * partition.
+     */
+    private static class AllReduceSend
+            extends RichFlatMapFunction<double[], Tuple3<Integer, Integer, double[]>> {
+        boolean hasReceivedOneRecord = false;
+        double[] transBuf = new double[TRANSFER_BUFFER_SIZE];
+
+        @Override
+        public void flatMap(
+                double[] receivedElement, Collector<Tuple3<Integer, Integer, double[]>> out) {
+            if (hasReceivedOneRecord) {
+                throw new RuntimeException("The input cannot contain more than one double array.");
+            }
+            hasReceivedOneRecord = true;
+            int numTasks = getRuntimeContext().getNumberOfParallelSubtasks();
+
+            for (int taskId = 0; taskId < numTasks; taskId++) {
+                int startChunkId = getStartChunkId(taskId, numTasks, receivedElement.length);
+                int numChunksToHandle =
+                        getNumChunksByTaskId(taskId, numTasks, receivedElement.length);
+                for (int chunk = 0; chunk < numChunksToHandle; chunk++) {
+                    int bufStart = (startChunkId + chunk) * TRANSFER_BUFFER_SIZE;
+                    System.arraycopy(
+                            receivedElement,
+                            bufStart,
+                            transBuf,
+                            0,
+                            getLengthOfChunk(startChunkId + chunk, receivedElement.length));
+                    out.collect(Tuple3.of(startChunkId + chunk, receivedElement.length, transBuf));
+                }
+            }
+        }
+    }
+
+    /** Aggregates partitioned array from other workers. */
+    private static class AllReduceSum
+            extends AbstractStreamOperator<Tuple4<Integer, Integer, Integer, double[]>>
+            implements OneInputStreamOperator<
+                            Tuple3<Integer, Integer, double[]>,
+                            Tuple4<Integer, Integer, Integer, double[]>>,
+                    BoundedOneInput {
+
+        /**
+         * A map that aggregates the received partial arrays. The key is chunkId, the value is
+         * (chunkId, totalLength, aggPartitionedArray).
+         */
+        Map<Integer, Tuple3<Integer, Integer, double[]>> sum = new HashMap<>();
```

Comment:
Maybe add the `private` keyword here? 

Can we simplify its type to be `Map<Integer, Tuple3<Integer, double[]>>`

And could we rename this variable to something more meaningful, e.g. `arrayByChunkId`?